### PR TITLE
CCSD

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**
In this PR, I will implement the CCSD quantum chemical method, together with new APIs that are required for it.

I will add Python bindings for CCSD later.

**Related issues**

**Additional context**

This PR features a refactor of `BlockMatrix` and `BlockRankFourTensor` into `ImplicitMatrixSlice` and `ImplicitRankFourTensorSlice`, generalizing their behavior.

It also introduces an `OrbitalSpace`, which encapsulates nicely sets of occupied, active and virtual orbitals.

**Don't forget:** 
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files
- [x] update the documentation on [our website](https://github.com/GQCG/gqcg.github.io) and check if new Python bindings should be added

### CODE REVIEW

File: gqcp/include/QCModel/HF/RHF.hpp [hunk 1]
```diff
@@ -207,21 +213,22 @@ class RHF {
...
         // Loop over all indices (ai,bj) to construct the orbital hessian
-        for (size_t a = N_P; a < K; a++) {
-            for (size_t i = 0; i < N_P; i++) {
+        for (const auto& a : orbital_space.indices(OccupationType::k_virtual)) {
+            for (const auto& i : orbital_space.indices(OccupationType::k_occupied)) {
 
-                for (size_t b = N_P; b < K; b++) {
-                    for (size_t j = 0; j < N_P; j++) {
+                for (const auto& b : orbital_space.indices(OccupationType::k_virtual)) {
+                    for (const auto& j : orbital_space.indices(OccupationType::k_occupied)) {
```
Comment (lelemmen) [inline gqcp/include/QCModel/HF/RHF.hpp:231]: @dariatols @xdvriend @guacke This is the API change that I'm most proud of. I think this is a really nice example of OOP doing its job with encapsulation, enhancing readability and making implementations less error-prone.

Comment (dariatols) [reply gqcp/include/QCModel/HF/RHF.hpp:231]: Awesome! Less error-prone and enhances the readability of the code!

Comment (dariatols) [reply gqcp/include/QCModel/HF/RHF.hpp:231]: Awesome! Now you can worry less about which indices are occupied/virtual. Nice!

**Comment by dariatols** [comment]
PR looks good to me! This code should help me a lot implementing CCD, thanks!

**Pull request merged**